nussl/separation/nmf_mfcc.py

```diff
@@ -0,0 +1,215 @@
+    """
+        Non Negative Matrix Factorization using K-Means Clustering on MFCC (NMF MFCC) is a source separation
+        algorithm that runs Transformer NMF on the magnitude spectrogram of an input audio signal.
+        It uses K means clustering to cluster the templates and activations returned by the NMF. The dot product of the
+        clustered templates and activations results in a magnitude spectrogram only containing a separated source.
+        This is used to create a Binary Mask object, and the whole process can be applied for each cluster to
+        return a list of Audio Signal objects corresponding to each separated source.
+
+        References:
+            Mel Frequency Cepstral Coefficients(MFCC): https://en.wikipedia.org/wiki/Mel-frequency_cepstrum
+            Volker Gnann, Martin Spiertz. "Source-filter based clustering for monaural blind source separation"
+            October 2002: https://www.researchgate.net/profile/Martin_Spiertz/publication/228567105_Source-filter_based
+            _clustering_for_monaural_blind_source_separation/links/0c960526b56dae0c69000000.pdf
+            scikit-learn KMeans: http://scikit-learn.org/stable/modules/generated/sklearn.cluster.KMeans.html
+
+        Parameters:
+            input_audio_signal (np.array): An Audio Signal object of the input audio signal.
+            num_sources (int): Number of sources to find.
+            num_templates (int): Number of template vectors to used in NMF.
+            distance_measure (str): The type of distance measure to use in NMF - euclidean or divergence.
+            num_iterations (int): The number of iterations to go through in NMF.
+            random_seed (int): The seed to use in the numpy random generator
+            kmeans_kwargs (dict): The kwargs for KMeans parameters.
+            convert_to_mono (bool): Given a stereo signal, convert to mono.
+            mask_type (object): A soft or binary mask object used for the source separation.
+            mfcc_range (int/list/tuple): The range of mfcc to use for clustering.
+            n_mfcc (int): The max number of mfccs to use.
+
+        Attributes:
+            input_audio_signal (object): An Audio Signal object of the input audio signal.
+            clusterer (object): A KMeans object for clustering the templates and activations.
+            signal_stft (np.matrix): The stft data for the input audio signal.
+            labeled_templates (np.array): A Numpy array containing the labeled templates columns
+                                               from the templates matrix for a particular source.
+            sources (np.array): A Numpy array containing the list of Audio Signal objects for each source.
+            masks (np.array): A Numpy array containing the lists of Binary Mask objects for each channel.
+
+        """
+    def __init__(self, input_audio_signal, num_sources, num_templates=50, distance_measure='euclidean',
```

Comment:
For `distance_measure=` use the `transformer_nmf.TransformerNMF.EUCLIDEAN` variable instead of the string literal. This is more verbose, but it has the advantage that everything is referencing a single value, so if it needs to be changed in the future, we won't have broken things all over the place.